File: src/rider/main/kotlin/com/jetbrains/rider/plugins/specflowriderplugin/SpecflowLanguageSubstitutor.java
package com.jetbrains.rider.plugins.specflowriderplugin;

import com.intellij.lang.Language;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.vfs.VfsUtilCore;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.openapi.vfs.VirtualFileVisitor;
import com.intellij.psi.LanguageSubstitutor;
import com.jetbrains.rider.plugins.specflowriderplugin.ideaInterop.fileTypes.GherkinLanguage;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Arrays;

public class SpecflowLanguageSubstitutor extends LanguageSubstitutor {
    @Override
    public @Nullable Language getLanguage(@NotNull VirtualFile virtualFile, @NotNull Project project) {
        // FIXME: not sure if this is call ofter or not, it may need some optimization or cache
        if (virtualFile.getName().endsWith(".feature")) {
            var parent = virtualFile.getParent();
            while (parent != null) {
                for (VirtualFile child : parent.getChildren()) {
                    if (child.getName().endsWith("proj") || child.getName().endsWith(".cs")) {
                        if (findCsharpFile(parent)) {
                            return GherkinLanguage.INSTANCE;
                        }
                        return null;
                    }
                }
                if (Arrays.stream(parent.getChildren()).anyMatch(f -> f.getName().endsWith(".sln"))) {
                    return null;
                }
                parent = parent.getParent();
            }
        }
        return null;
    }

    private boolean findCsharpFile(VirtualFile directory) {
        final boolean[] found = {false};
        VfsUtilCore.visitChildrenRecursively(directory, new VirtualFileVisitor<Object>() {
            @Override
            public boolean visitFile(@NotNull VirtualFile file) {
                if (found[0])
                    return false;
                if (file.getName().endsWith(".cs")) {
                    found[0] = true;
                    return false;
                }
                return super.visitFile(file);
            }
        });
        return found[0];
    }
}

File: src/rider/main/kotlin/com/jetbrains/rider/plugins/specflowriderplugin/SpecflowLanguageSubstitutor.kt
package com.jetbrains.rider.plugins.specflowriderplugin

import com.intellij.lang.Language
import com.intellij.openapi.project.Project
import com.intellij.psi.LanguageSubstitutor
import com.intellij.openapi.vfs.VirtualFile
import com.jetbrains.rider.plugins.specflowriderplugin.ideaInterop.fileTypes.GherkinLanguage
import com.intellij.openapi.vfs.VfsUtilCore
import com.intellij.openapi.vfs.VirtualFileVisitor
import java.util.*

class SpecflowLanguageSubstitutor : LanguageSubstitutor() {
    override fun getLanguage(virtualFile: VirtualFile, project: Project): Language? {
        // FIXME: not sure if this is call ofter or not, it may need some optimization or cache
        if (virtualFile.name.endsWith(".feature")) {
            var parent = virtualFile.parent
            while (parent != null) {
                for (child in parent.children) {
                    if (child.name.endsWith("proj") || child.name.endsWith(".cs")) {
                        return if (findCsharpFile(parent)) {
                            GherkinLanguage
                        } else null
                    }
                }
                if (Arrays.stream(parent.children).anyMatch { f: VirtualFile -> f.name.endsWith(".sln") }) {
                    return null
                }
                parent = parent.parent
            }
        }
        return null
    }

    private fun findCsharpFile(directory: VirtualFile): Boolean {
        val found = booleanArrayOf(false)
        VfsUtilCore.visitChildrenRecursively(directory, object : VirtualFileVisitor<Any?>() {
            override fun visitFile(file: VirtualFile): Boolean {
                if (found[0]) return false
                if (file.name.endsWith(".cs")) {
                    found[0] = true
                    return false
                }
                return super.visitFile(file)
            }
        })
        return found[0]
    }
}